### PR TITLE
[PT-168073522] [aepp-demo] SC reconnect more flexible

### PR DESCRIPTION
**Pull request by uwiger**
See PR #2758 - commits squashed and cherry-picked onto the `aeuniverse-demo` branch.

Test suites pass locally. State password given a default value in the `sc_ws_handler` options parsing.

### CODE REVIEW

File: apps/aechannel/src/aesc_fsm.hrl [hunk 1]
```diff
@@ -256,7 +263,8 @@
                   , data    :: #op_data{}
                   }).
 
--record(op_reestablish, {offchain_tx :: aetx_sign:signed_tx()
+-record(op_reestablish, { offchain_tx :: aetx_sign:signed_tx() | undefined
```
Comment (tolbrino) [inline apps/aechannel/src/aesc_fsm.hrl]: The semantic of both modes should be documented here.

File: apps/aehttp/src/sc_ws_handler.erl [hunk 2]
```diff
@@ -52,46 +85,125 @@ websocket_init(Params) ->
...
+    case start_link_fsm(Handler, ChannelOpts) of
+        {ok, FsmPid} ->
+            MRef = erlang:monitor(process, FsmPid),
+            {ok, Handler#handler{fsm_pid = FsmPid, fsm_mref = MRef}};
+        {error, Err} ->
+            handler_init_error(Err, Handler)
     end.
 
-websocket_init_reconnect(#{ <<"reconnect_tx">> := ReconnectTx } = Params) ->
+websocket_init_reconnect(#{ job_id := JobId
```
Comment (tolbrino) [inline apps/aehttp/src/sc_ws_handler.erl]: I'd like to see documentation here explaining the chain of logic which is followed to try and reconnect.

File: apps/aehttp/src/sc_ws_handler.erl [hunk 3]
```diff
@@ -52,46 +85,125 @@ websocket_init(Params) ->
...
+            OnError(not_found)
+    end.
+
+check_existing_client(Client, Opts, Handler) ->
+    OnError = fun(E) ->
+                      handler_init_error(E, Handler)
+              end,
+    T = time_since_last_dispatch(Client),
+    lager:debug("Time since last dispatch (~p): ~p", [Client, T]),
+    if is_integer(T), T < 5000 ->
```
Comment (tolbrino) [inline apps/aehttp/src/sc_ws_handler.erl]: The timeout should be documented and configurable. The configuration can be in a different PR though.

File: apps/aehttp/src/sc_ws_handler.erl [hunk 4]
```diff
@@ -308,6 +426,18 @@ prepare_handler(#{<<"protocol">> := Protocol} = Params) ->
...
 
+update_handler_for_reestablish(Handler, Opts) ->
+    lager:debug("Handler = ~p; Opts = ~p", [lager:pr(Handler, ?MODULE), Opts]),
+    Conn = maps:get(connection, Opts),
+    Handler#handler{
+       role = maps:get(role, Opts)
+     , host = maps:get(host, Conn, undefined)
+     , port = maps:get(port, Conn) }.
+
+expand_cached_opts(Opts) ->
```
Comment (tolbrino) [inline apps/aehttp/src/sc_ws_handler.erl]: It should be documented which parameters overwrite which parameters.

File: apps/aechannel/src/aesc_fsm.erl [hunk 5]
```diff
@@ -382,6 +384,18 @@ error_code_to_msg(Code) when Code >= ?ERR_USER ->
 error_code_to_msg(_) ->
     <<"unknown">>.
 
+-spec stop(pid()) -> ok.
+stop(Fsm) ->
+    lager:debug("Gracefully stopping the FSM ~p", [Fsm]),
+    stop_ok(catch gen_statem:stop(Fsm, shutdown, 5000)).
+
+stop_ok(ok) ->
```
Comment (tolbrino) [inline apps/aechannel/src/aesc_fsm.erl]: Can you document why these clauses are handled specifically as ok?

File: apps/aechannel/src/aesc_fsm.erl [hunk 6]
```diff
@@ -3602,32 +3714,60 @@ check_block_hash_deltas(#{block_hash_delta := InvalidBHDelta} = Opts) ->
...
+
 check_opts([], Opts) ->
     Opts;
 check_opts([H|T], Opts) ->
     check_opts(T, H(Opts)).
 
+prepare_for_reestablish(#data{ opts = Opts
+                             , on_chain_id = ChanId } = D) ->
+    try
+        Session = start_session(#{ existing_channel_id => ChanId },
```
Comment (grzegorz225) [inline apps/aechannel/src/aesc_fsm.erl]: Make white-spaces inside maps consistent.

File: apps/aechannel/src/aesc_state_cache.erl [hunk 7]
```diff
@@ -197,7 +200,7 @@ transform_record_to_encrypted_form_with_default_password(#pch{id = ChId, pubkeys
     [encrypt_cache(Cache1)];
 transform_record_to_encrypted_form_with_default_password(#pch{pubkeys = [Pub1, Pub2]} = Record) ->
     Records = [Record#pch{pubkeys = [Pub1]}, Record#pch{pubkeys = [Pub2]}],
-    lists:flatten([transform_record_to_encrypted_form_with_default_password(Record) || Record <- Records]).
+    lists:flatten([transform_record_to_encrypted_form_with_default_password(R) || R <- Records]).
```
Comment (grzegorz225) [inline apps/aechannel/src/aesc_state_cache.erl]: The migrated records will contain an undefined opts map. We need to create a opts map here.

File: apps/aechannel/test/aesc_fsm_SUITE.erl [hunk 8]
```diff
@@ -1159,48 +1190,102 @@ t_leave_reestablish_(Cfg) ->
 
 leave_reestablish_(Cfg) ->
     Debug = get_debug(Cfg),
-    {I0, R0, Spec0} = channel_spec(Cfg),
-    #{ i := #{} = I
-     , r := #{} = R
+    {I0, R0, Spec0} = InitialSpec = channel_spec(Cfg),
+    ?LOG(Debug, "Initial spec = ~p", [InitialSpec]),
+    #{ i := #{} = I01
```
Comment (grzegorz225) [inline apps/aechannel/test/aesc_fsm_SUITE.erl]: Can you use sequential numbers here? So I0, I1, I2 ... instead of I0, I01

File: apps/aechannel/test/aesc_fsm_SUITE.erl [hunk 9]
```diff
@@ -1159,48 +1190,102 @@ t_leave_reestablish_(Cfg) ->
...
+    ?PEEK_MSGQ,
+    FsmStateI = check_fsm_state(FsmI),
+    FsmStateR = check_fsm_state(FsmR),
+    {FsmStateI, FsmStateR} = {FsmStateR, FsmStateI},
+    ?LOG(Debug, "verifying channel: update", []),
+    {I1, R1} = do_update(PubI, PubR, 1, I, R, Debug, Cfg),
+    ?PEEK_MSGQ,
+    ?LOG(Debug, "verifying channel: deposit", []),
+    {ok, I2, R2} = deposit_(I1, R1, 1, Debug, Cfg),
+    ?LOG("=== Channel verified~nMessage Q: ~p", [element(2,process_info(self(), messages))]),
```
Comment (grzegorz225) [inline apps/aechannel/test/aesc_fsm_SUITE.erl]: Why not use the PEEK_MSGQ macro here?

File: apps/aechannel/test/aesc_fsm_SUITE.erl [hunk 10]
```diff
@@ -2302,32 +2406,37 @@ create_channel_from_spec(I, R, Spec, Port, UseAny, Debug, Cfg) ->
     %% TODO: Somehow there is a CI only race condition which rarely occurs in
     %% round_too_high.check_incorrect_* and round_too_low.check_incorrect_* tests
     %% For now just wrap this operation in a retry loop and come back to it later
+    ?LOG("=== Create channel~n"
+         "Spec = ~p"
+         "Msg Q: ~p", [Spec, element(2, process_info(self(), messages))]),
```
Comment (grzegorz225) [inline apps/aechannel/test/aesc_fsm_SUITE.erl]: Can we use the PEEK_MSGQ macro here?

File: apps/aechannel/test/aesc_fsm_SUITE.erl [hunk 11]
```diff
@@ -2351,22 +2460,34 @@ create_channel_from_spec(I, R, Spec, Port, UseAny, Debug, Cfg) ->
...
+custom_spec_opts(responder, Cfg) ->
+    proplists:get_value(responder_opts, Cfg, #{});
+custom_spec_opts(initiator, Cfg) ->
+    proplists:get_value(initiator_opts, Cfg, #{}).
 
 spawn_responder(Port, Spec, R, UseAny, Debug) ->
     Me = self(),
-    spawn(fun() ->
-                       log("responder spawned: ~p", [Spec]),
+    spawn_link(fun() ->
```
Comment (grzegorz225) [inline apps/aechannel/test/aesc_fsm_SUITE.erl]: In case a rare CI crash occurs in the child process the crash will kill the test process instead of retrying the operation :)

File: apps/aechannel/test/aesc_fsm_SUITE.erl [hunk 12]
```diff
@@ -2764,27 +2910,30 @@ receive_from_fsm(Tag, R, Msg, Timeout, Debug, Cont) ->
...
-                    log(Debug, "Failed match: ~p", [Msg1]),
+                    ?LOG(Debug, "Failed match: ~p", [Msg1]),
                     receive_from_fsm_(Tag, R, Msg, TRef, Debug, Cont)
             after
                 erlang:cancel_timer(TRef)
             end;
         {timeout, TRef, receive_from_fsm} ->
             flush(),
             error(timeout)
+    after ?TIMEOUT ->
```
Comment (grzegorz225) [inline apps/aechannel/test/aesc_fsm_SUITE.erl]: Use the PEEK_MSGQ macro here.

File: apps/aehttp/src/sc_ws_handler.erl [hunk 13]
```diff
@@ -52,46 +85,125 @@ websocket_init(Params) ->
...
-                    case aesc_fsm:reconnect_client(Fsm, self(), SignedTx) of
-                        ok ->
-                            MRef = erlang:monitor(process, Fsm),
-                            { ok, Handler#handler{ fsm_pid  = Fsm
-                                                 , fsm_mref = MRef } };
-                        {error, Err} ->
-                            handler_init_error(Err, Handler)
-                    end
+        {ok, #{} = Opts} ->
+            lager:debug("Reconnect tx Opts = ~p", [Opts]),
```
Comment (grzegorz225) [inline apps/aehttp/src/sc_ws_handler.erl]: This will leak the current user provided state password :) The protection against leakage kicks in only in the FSM process.

File: apps/aehttp/test/aehttp_sc_SUITE.erl [hunk 14]
```diff
@@ -3058,11 +3125,53 @@ sc_ws_cheating_close_solo_(Config, ChId, Poi, WhoCloses) ->
...
-
+    ?CHECK_INFO(20),
     Config1 = sc_ws_reestablish_(ReestablishOptions, Config),
     ok = sc_ws_update_(Config1),
     ok = sc_ws_close_(Config1).
 
+sc_ws_leave_reestablish_responder_stays(Config0) ->
+    Config = sc_ws_open_(Config0, #{responder_opts => #{keep_running => true}}),
+    Config1 = [{responder_leaves, false}|Config],
+    ReestablOptions = sc_ws_leave_(Config1),
```
Comment (grzegorz225) [inline apps/aehttp/test/aehttp_sc_SUITE.erl]: Typo in variable name

File: apps/aehttp/test/aehttp_sc_SUITE.erl [hunk 15]
```diff
@@ -3058,11 +3125,53 @@ sc_ws_cheating_close_solo_(Config, ChId, Poi, WhoCloses) ->
...
+
+sc_ws_leave_reconnect(Config0) ->
+    Config = sc_ws_open_(Config0, #{responder_opts => #{keep_running => true}}),
+    Participants = proplists:get_value(participants, Config),
+    #{ pub_key := Pub, priv_key := Priv } = maps:get(initiator, Participants),
+    ct:log("*** Leaving channel ***", []),
+    Config1 = [{responder_leaves, false}|Config],
+    #{ existing_channel_id := ChId
+     , offchain_tx         := _Tx }
+        = ReestOpts = sc_ws_leave_(Config1),
```
Comment (grzegorz225) [inline apps/aehttp/test/aehttp_sc_SUITE.erl]: ReestOpts -> ReestablishOpts

File: apps/aechannel/src/aesc_fsm.hrl [hunk 16]
```diff
@@ -150,12 +156,13 @@
...
               , cur_statem_state                :: undefined | atom()
               , state                           :: aesc_offchain_state:state() | function()
-              , session                         :: pid()
+              , session                         :: pid() | undefined
               , client                          :: pid() | undefined
               , client_mref                     :: reference() | undefined
               , client_connected = true         :: boolean()
               , client_may_disconnect = false   :: boolean()
               , client_reconnect_nonce = 0      :: non_neg_integer()
+              , peer_connected = false        :: boolean()
```
Comment (tolbrino) [inline apps/aechannel/src/aesc_fsm.hrl]:
```suggestion
              , peer_connected = false          :: boolean()
```

File: apps/aechannel/src/aesc_session_noise.erl [hunk 17]
```diff
@@ -337,7 +342,15 @@ code_change(_FromVsn, St, _Extra) ->
     {ok, St}.
 
 locate_fsm(Type, MInfo, #st{ init_op = {accept, SnInfo, _Opts} } = St) ->
-    Info = maps:merge(SnInfo, MInfo),  % any duplicates overridden by what was received
+    lager:debug("Type = ~p, MInfo = ~p, SnInfo = ~p", [Type, MInfo, SnInfo]),
+    %% any duplicates overridden by what was received
+    Info0 = maps:merge(maps:with([ initiator
+                                 , responder
+                                 , port ], SnInfo), MInfo),
+    Info = case Type of
```
Comment (tolbrino) [inline apps/aechannel/src/aesc_session_noise.erl]: Info = Info0#{reestablish => (Type =:= ?CH_REESTABL)},

File: apps/aechannel/src/aesc_state_cache.erl [hunk 18]
```diff
@@ -78,8 +78,10 @@
              min_depth = ?MIN_DEPTH}).
 -type ch_cache_id() :: { aeser_id:val()
                        , aeser_id:val() | atom()}.
+-type opts() :: map().
 -record(ch_cache, { cache_id    :: ch_cache_id() | undefined
                   , state       :: aesc_offchain_state:state() | atom() | undefined
+                  , opts = #{}  :: map() | '_'
```
Comment (tolbrino) [inline apps/aechannel/src/aesc_state_cache.erl]: Why the catch all?

File: apps/aechannel/src/aesc_state_cache.erl [hunk 19]
```diff
@@ -91,26 +93,27 @@
...
 -define(SERVER, ?MODULE).
 -define(TAB, aesc_state_cache_ch).
 -define(PTAB, aesc_state_cache).
 -define(CACHE_DEFAULT_PASSWORD, "correct horse battery staple").
 
--spec new(aeser_id:val(), aeser_id:val(), aesc_offchain_state:state()) -> ok | {error, any()}.
-new(ChId, Pubkey, State) ->
-    new(ChId, Pubkey, State, ?CACHE_DEFAULT_PASSWORD).
+-spec new(aeser_id:val(), aeser_id:val(), aesc_offchain_state:state(), map()) -> ok | {error, any()}.
+new(ChId, Pubkey, State, Opts) ->
```
Comment (tolbrino) [inline apps/aechannel/src/aesc_state_cache.erl]: Having the type and the parameter which doesn't use the type is confusing.

File: apps/aechannel/src/aesc_fsm.erl (hunk 5 above)
Comment (uwiger) [reply apps/aechannel/src/aesc_fsm.erl]: This function should be removed entirely as soon as it's removed from the `aehttp_sc_SUITE`

File: apps/aehttp/test/aehttp_ws_test_utils.erl [hunk 20]
```diff
@@ -60,7 +60,7 @@
                }).
 
 start_link(Host, Port) ->
-    WsAddress = "ws://" ++ Host ++ ":" ++ integer_to_list(Port) ++ "/websocket",
+    WsAddress = "wss://" ++ Host ++ ":" ++ integer_to_list(Port) ++ "/websocket",
```
Comment (tolbrino) [inline apps/aehttp/test/aehttp_ws_test_utils.erl]: Can you add a TODO comment here to make this configurable?

File: apps/aechannel/src/aesc_state_cache.erl (hunk 18 above)
Comment (uwiger) [reply apps/aechannel/src/aesc_state_cache.erl]: It's used in a select pattern, and Dialyzer gets cranky otherwise. :)

File: apps/aechannel/src/aesc_state_cache.erl (hunk 7 above)
Comment (uwiger) [reply apps/aechannel/src/aesc_state_cache.erl]: The default value for `opts` in `#ch_cache{}` is `#{}`...

File: apps/aechannel/test/aesc_fsm_SUITE.erl (hunk 8 above)
Comment (uwiger) [reply apps/aechannel/test/aesc_fsm_SUITE.erl]: I think it should be refactored into smaller functions instead.

File: apps/aechannel/src/aesc_fsm.erl [hunk 21]
```diff
@@ -2652,6 +2728,22 @@ check_inband_msg(#{ channel_id := ChanId
 check_inband_msg(_, _) ->
     {error, chain_id_mismatch}.
 
+check_client_reconnect_tx(SignedTx, #data{ opts = #{ reconnect_security := none } } = D) ->
```
Comment (grzegorz225) [inline apps/aechannel/src/aesc_fsm.erl]: Shouldn't this parameter be documented somewhere?

Comment (uwiger) [reply apps/aechannel/src/aesc_fsm.erl]: Not sure if we want to keep it in its current form after the conference. :)
I would prefer a reconnect check based on a session token instead.

File: apps/aechannel/src/aesc_fsm.erl [hunk 22]
```diff
@@ -553,48 +590,53 @@ awaiting_open({call, From}, {attach_responder, Info}, #data{opts = Opts} = D) ->
...
-         , gproc_key  := K } ->
+    case check_attach_info(Info, D) of
+        ok ->
             gproc:unreg(K),
             gen:reply(From, ok),
             {Pid, _} = From,
-            keep_state(D#data{session = Pid}, [{reply, From, ok}]);
-        _ ->
+            keep_state(D#data{ session = Pid }, [{reply, From, ok}]);
+        {error,_} ->
```
Comment (velzevur) [inline apps/aechannel/src/aesc_fsm.erl]:
```suggestion
        {error, Error} ->
                lager:debug("Attach failed with ~p", [Error]),
```

File: apps/aechannel/src/aesc_fsm.erl [hunk 23]
```diff
@@ -2949,15 +3041,29 @@ on_chain_id(D, SignedTx) ->
...
-            aesc_state_cache:new(ChId, my_account(D), State)
+            aesc_state_cache:new(ChId, my_account(D), State, Opts)
     end,
     D#data{state_password_wrapper = undefined}.
 
+opts_to_cache() ->
+    [ role
+    , initiator
+    , responder
+    , connection
```
Comment (velzevur) [inline apps/aechannel/src/aesc_fsm.erl]: I see `connection` holds the `host` and `port` from opening params

File: apps/aehttp/src/sc_ws_handler.erl [hunk 24]
```diff
@@ -50,48 +83,160 @@ websocket_init(Params) ->
...
-                        ok ->
-                            MRef = erlang:monitor(process, Fsm),
-                            { ok, Handler#handler{ fsm_pid  = Fsm
-                                                 , fsm_mref = MRef } };
-                        {error, Err} ->
-                            handler_init_error(Err, Handler)
-                    end
-            end;
+        {ok, #{} = Opts} ->
+            lager:debug("Reconnect tx Opts = ~p", [Opts]),
```
Comment (tolbrino) [inline apps/aehttp/src/sc_ws_handler.erl]:
```suggestion
            lager:debug("Reconnect tx Opts = ~p", [aesc_utils:censor_init_opts(Opts)]),
```

File: apps/aechannel/test/aesc_fsm_SUITE.erl (hunk 9 above)
Comment (uwiger) [inline apps/aechannel/test/aesc_fsm_SUITE.erl]:
```suggestion
    ?LOG("=== Channel verified", []),
    ?PEEK_MSGQ,
```

File: apps/aechannel/test/aesc_fsm_SUITE.erl [hunk 25]
```diff
@@ -2302,32 +2406,37 @@ create_channel_from_spec(I, R, Spec, Port, UseAny, Debug, Cfg) ->
     %% TODO: Somehow there is a CI only race condition which rarely occurs in
     %% round_too_high.check_incorrect_* and round_too_low.check_incorrect_* tests
     %% For now just wrap this operation in a retry loop and come back to it later
+    ?LOG("=== Create channel~n"
+         "Spec = ~p"
```
Comment (uwiger) [inline apps/aechannel/test/aesc_fsm_SUITE.erl]:
```suggestion
    ?LOG("Spec = ~p", [Spec]),
```

File: apps/aechannel/test/aesc_fsm_SUITE.erl (hunk 10 above)
Comment (uwiger) [inline apps/aechannel/test/aesc_fsm_SUITE.erl]:
```suggestion
    ?PEEK_MSGQ,
```

File: apps/aechannel/src/aesc_state_cache.erl [hunk 26]
```diff
@@ -91,26 +93,27 @@
...
 
 -define(SERVER, ?MODULE).
 -define(TAB, aesc_state_cache_ch).
 -define(PTAB, aesc_state_cache).
 -define(CACHE_DEFAULT_PASSWORD, "correct horse battery staple").
 
--spec new(aeser_id:val(), aeser_id:val(), aesc_offchain_state:state()) -> ok | {error, any()}.
-new(ChId, Pubkey, State) ->
-    new(ChId, Pubkey, State, ?CACHE_DEFAULT_PASSWORD).
+-spec new(aeser_id:val(), aeser_id:val(), aesc_offchain_state:state(), map()) -> ok | {error, any()}.
```
Comment (velzevur) [inline apps/aechannel/src/aesc_state_cache.erl]:
```suggestion
-spec new(aeser_id:val(), aeser_id:val(), aesc_offchain_state:state(), opts()) -> ok | {error, any()}.
```

File: apps/aechannel/src/aesc_state_cache.erl [hunk 27]
```diff
@@ -91,26 +93,27 @@
...
+new(ChId, Pubkey, State, Opts) ->
+    new(ChId, Pubkey, State, Opts, ?CACHE_DEFAULT_PASSWORD).
 
--spec new(aeser_id:val(), aeser_id:val(), aesc_offchain_state:state(), string()) -> ok | {error, any()}.
-new(ChId, PubKey, State, Password) ->
-    gen_server:call(?SERVER, {new, ChId, PubKey, self(), State, unicode:characters_to_binary(Password)}).
+-spec new(aeser_id:val(), aeser_id:val(), aesc_offchain_state:state(), map(), string()) -> ok | {error, any()}.
```
Comment (velzevur) [inline apps/aechannel/src/aesc_state_cache.erl]:
```suggestion
-spec new(aeser_id:val(), aeser_id:val(), aesc_offchain_state:state(), opts(), string()) -> ok | {error, any()}.
```

File: apps/aechannel/src/aesc_session_noise.erl (hunk 17 above)
Comment (uwiger) [inline apps/aechannel/src/aesc_session_noise.erl]:
```suggestion
    Info = Info0#{reestablish => (Type =:= ?CH_REESTABL)},
```

File: apps/aechannel/src/aesc_session_noise.erl [hunk 28]
```diff
@@ -337,7 +342,15 @@ code_change(_FromVsn, St, _Extra) ->
...
+               ?CH_OPEN     -> Info0;
```
Comment (uwiger) [inline apps/aechannel/src/aesc_session_noise.erl]:
```suggestion
```

File: apps/aechannel/src/aesc_session_noise.erl [hunk 29]
```diff
@@ -337,7 +342,15 @@ code_change(_FromVsn, St, _Extra) ->
...
+               ?CH_REESTABL -> Info0#{reestablish => true}
```
Comment (uwiger) [inline apps/aechannel/src/aesc_session_noise.erl]:
```suggestion
```

File: apps/aechannel/src/aesc_session_noise.erl [hunk 30]
```diff
@@ -337,7 +342,15 @@ code_change(_FromVsn, St, _Extra) ->
...
+           end,
```
Comment (uwiger) [inline apps/aechannel/src/aesc_session_noise.erl]:
```suggestion
```

File: apps/aehttp/test/aehttp_ws_test_utils.erl (hunk 20 above)
Comment (uwiger) [inline apps/aehttp/test/aehttp_ws_test_utils.erl]:
```suggestion
    % TODO: Make it configurable whether to use `ws` or `wss`
    WsAddress = "wss://" ++ Host ++ ":" ++ integer_to_list(Port) ++ "/websocket",
```

File: apps/aechannel/test/aesc_fsm_SUITE.erl [hunk 31]
```diff
@@ -2271,14 +2375,15 @@ channel_spec(Cfg, ChannelReserve, PushAmount, XOpts) ->
     {I1, R1, Spec1}.
 
 prime_password(#{} = P, Key, Cfg) when Key =:= initiator_password; Key =:= responder_password ->
+    Debug = get_debug(Cfg),
     case ?config(Key, Cfg) of
         undefined ->
             P#{state_password => generate_password()};
         ignore ->
-            ct:log("Ignoring password"),
+            ?LOG(Debug, "Ignoring password"),
```
Comment (velzevur) [inline apps/aechannel/test/aesc_fsm_SUITE.erl]:
```suggestion
            ?LOG(Debug, "Ignoring password", []),
```

File: apps/aechannel/test/aesc_fsm_SUITE.erl [hunk 32]
```diff
@@ -2271,14 +2375,15 @@ channel_spec(Cfg, ChannelReserve, PushAmount, XOpts) ->
...
             maps:remove(state_password, P);
         Password ->
-            ct:log("Using predefined password"),
+            ?LOG(Debug, "Using predefined password"),
```
Comment (velzevur) [inline apps/aechannel/test/aesc_fsm_SUITE.erl]:
```suggestion
            ?LOG(Debug, "Using predefined password", []),
```

File: apps/aehttp/src/sc_ws_handler.erl (hunk 3 above)
Comment (velzevur) [reply apps/aehttp/src/sc_ws_handler.erl]: This PR has enough stuff already...
